File: finding-seams/kotlin/src/main/kotlin/findingseams/badgeprinting/BadgePrinting.kt
package findingseams.badgeprinting

import java.time.Instant

class BadgePrinting(
    private val printerDrivers: PrinterDrivers? = null,
    private val printerConfigs: PrinterConfigs? = null,
    private val badgeDimensionResolvers: BadgeDimensionResolvers? = null,
    private val printerFactory: PrinterFactory? = null,
) {
    fun createBadge(name: String, pronouns: String): Badge {
        val issuedAt = Instant.now()
        val id = BadgeSequence.nextId()
        val formattedId = "CONF-$id"
        return Badge(formattedId, name, pronouns, issuedAt)
    }

    @Suppress("UNUSED_PARAMETER", "UNCHECKED_CAST")
    fun printBadge(badge: Badge) {
        val resolvers = badgeDimensionResolvers as BadgeDimensionResolvers
        val configs = printerConfigs as PrinterConfigs
        val drivers = printerDrivers as PrinterDrivers
        val factory = printerFactory as PrinterFactory
        val dimensions = resolvers.resolve(badge)
        val config = configs.getConfigBasedOnDimensions(dimensions)
        val driver = drivers.get("BadgePrinting")
        val printer = factory.load(driver, config)
        printer.print(badge)
    }
}

interface BadgeDimensionResolvers {
    fun resolve(badge: Badge): Any
}

interface PrinterConfigs {
    fun getConfigBasedOnDimensions(dimensions: Any): Any
}

interface PrinterDrivers {
    fun get(name: String): Any
}

interface PrinterFactory {
    fun load(driver: Any, config: Any): Printer
}

interface Printer {
    fun print(badge: Badge)
}

File: finding-seams/csharp/FindingSeams/BadgePrinting/BadgePrinting.cs
namespace FindingSeams.BadgePrinting;

public class BadgePrinting
{
    private readonly object? _printerDrivers;
    private readonly object? _printerConfigs;
    private readonly object? _badgeDimensionResolvers;
    private readonly object? _printerFactory;

    public BadgePrinting(
        object? printerDrivers = null,
        object? printerConfigs = null,
        object? badgeDimensionResolvers = null,
        object? printerFactory = null)
    {
        _printerDrivers = printerDrivers;
        _printerConfigs = printerConfigs;
        _badgeDimensionResolvers = badgeDimensionResolvers;
        _printerFactory = printerFactory;
    }

    public Badge CreateBadge(string name, string pronouns)
    {
        var issuedAt = DateTimeOffset.Now;
        var id = BadgeSequence.NextId();
        var formattedId = $"CONF-{id}";
        return new Badge(formattedId, name, pronouns, issuedAt);
    }

    public void PrintBadge(Badge badge)
    {
        dynamic resolvers = _badgeDimensionResolvers!;
        dynamic configs = _printerConfigs!;
        dynamic drivers = _printerDrivers!;
        dynamic factory = _printerFactory!;
        dynamic dimensions = resolvers.Resolve(badge);
        dynamic config = configs.GetConfigBasedOnDimensions(dimensions);
        dynamic driver = drivers.Get("BadgePrinting");
        dynamic printer = factory.Load(driver, config);
        printer.Print(badge);
    }
}
